x-pack/filebeat/input/netflow/definitions.go

```diff
@@ -0,0 +1,217 @@
+
+import (
+	"fmt"
+	"io/ioutil"
+	"math"
+	"os"
+	"strconv"
+
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+
+	"github.com/elastic/beats/x-pack/filebeat/input/netflow/decoder/fields"
+)
+
+var logstashName2Decoder = map[string]fields.Decoder{
+	"double":                  fields.Float64,
+	"float":                   fields.Float32,
+	"int8":                    fields.Signed8,
+	"int15":                   fields.SignedDecoder(15),
+	"int16":                   fields.Signed16,
+	"int24":                   fields.SignedDecoder(24),
+	"int32":                   fields.Signed32,
+	"int64":                   fields.Signed64,
+	"ip4_addr":                fields.Ipv4Address,
+	"ip4addr":                 fields.Ipv4Address,
+	"ip6_addr":                fields.Ipv6Address,
+	"ip6addr":                 fields.Ipv6Address,
+	"mac_addr":                fields.MacAddress,
+	"macaddr":                 fields.MacAddress,
+	"string":                  fields.String,
+	"uint8":                   fields.Unsigned8,
+	"uint15":                  fields.UnsignedDecoder(15),
+	"uint16":                  fields.Unsigned16,
+	"uint24":                  fields.UnsignedDecoder(24),
+	"uint32":                  fields.Unsigned32,
+	"uint64":                  fields.Unsigned64,
+	"octet_array":             fields.OctetArray,
+	"octetarray":              fields.OctetArray,
+	"acl_id_asa":              fields.UnsupportedDecoder{},
+	"mpls_label_stack_octets": fields.UnsupportedDecoder{},
+	"application_id":          fields.UnsupportedDecoder{},
+	"forwarding_status":       fields.UnsupportedDecoder{},
+}
+
+// LoadFieldDefinitions takes a parsed YAML tree from a Logstash
+// Netflow or IPFIX custom fields format and converts it to a FieldDict.
+func LoadFieldDefinitions(yaml interface{}) (defs fields.FieldDict, err error) {
+	tree, ok := yaml.(map[interface{}]interface{})
+	if !ok {
+		return nil, fmt.Errorf("invalid custom fields definition format: expected a mapping of integer keys. Got %T", yaml)
+	}
+	if len(tree) == 0 {
+		return nil, nil
+	}
+	isIPFIX, err := fieldsAreIPFIX(tree)
+	if err != nil {
+		return nil, err
+	}
+	defs = fields.FieldDict{}
+	if !isIPFIX {
+		if err := loadFields(tree, 0, defs); err != nil {
+			return nil, errors.Wrap(err, "failed to load NetFlow fields")
+		}
+		return defs, nil
+	}
+	for pemI, fields := range tree {
+		pem, err := toInt(pemI)
+		if err != nil {
+			return nil, err
+		}
+		if !fits(pem, 0, math.MaxUint32) {
+			return nil, fmt.Errorf("PEM %d out of uint32 range", pem)
+		}
+		tree, ok := fields.(map[interface{}]interface{})
+		if !ok {
+			return nil, fmt.Errorf("IPFIX fields for pem=%d malformed", pem)
+		}
+		if err := loadFields(tree, uint32(pem), defs); err != nil {
+			return nil, errors.Wrapf(err, "failed to load IPFIX fields for pem=%d", pem)
+		}
+	}
+	return defs, nil
+}
+
+// LoadFieldDefinitionsFromFile takes the path to a YAML file in Logstash
+// Netflow or IPFIX custom fields format and converts it to a FieldDict.
```

Comment:
We'll need a description of the config format in the Filebeat docs.

Reply:
Done